Gems/ROS2/Code/Source/Camera/CameraSensor.cpp

```diff
@@ -0,0 +1,240 @@
+
+        const AZStd::string pipelineName = m_cameraSensorDescription.m_cameraName + "Pipeline" + getPipelineTypeName();
+
+        AZ::RPI::RenderPipelineDescriptor pipelineDesc;
+        pipelineDesc.m_mainViewTagName = "MainCamera";
+        pipelineDesc.m_name = pipelineName;
+
+        pipelineDesc.m_rootPassTemplate = getPipelineTemplateName();
+
+        pipelineDesc.m_renderSettings.m_multisampleState = AZ::RPI::RPISystemInterface::Get()->GetApplicationMultisampleState();
+        m_pipeline = AZ::RPI::RenderPipeline::CreateRenderPipeline(pipelineDesc);
+        m_pipeline->RemoveFromRenderTick();
+
+        if (auto renderToTexturePass = azrtti_cast<AZ::RPI::RenderToTexturePass*>(m_pipeline->GetRootPass().get()))
+        {
+            renderToTexturePass->ResizeOutput(m_cameraSensorDescription.m_width, m_cameraSensorDescription.m_height);
+        }
+
+        m_scene->AddRenderPipeline(m_pipeline);
+
+        m_passHierarchy.push_back(pipelineName);
+        m_passHierarchy.push_back("CopyToSwapChain");
+
+        m_pipeline->SetDefaultView(m_view);
+        const AZ::RPI::ViewPtr targetView = m_scene->GetDefaultRenderPipeline()->GetDefaultView();
+        if (auto* fp = m_scene->GetFeatureProcessor<AZ::Render::PostProcessFeatureProcessor>())
+        {
+            fp->SetViewAlias(m_view, targetView);
+        }
+    }
+
+    CameraSensor::~CameraSensor()
+    {
+        if (m_scene)
+        {
+            if (auto* fp = m_scene->GetFeatureProcessor<AZ::Render::PostProcessFeatureProcessor>())
+            {
+                fp->RemoveViewAlias(m_view);
+            }
+            m_scene->RemoveRenderPipeline(m_pipeline->GetId());
+            m_scene = nullptr;
+        }
+        m_passHierarchy.clear();
+        m_pipeline.reset();
+        m_view.reset();
+    }
+
+    void CameraSensor::RequestFrame(
+        const AZ::Transform& cameraPose, AZStd::function<void(const AZ::RPI::AttachmentReadback::ReadbackResult& result)> callback)
+    {
+        const AZ::Transform inverse = (cameraPose * kAtomToRos).GetInverse();
+        m_view->SetWorldToViewMatrix(AZ::Matrix4x4::CreateFromQuaternionAndTranslation(inverse.GetRotation(), inverse.GetTranslation()));
+
+        AZ::Render::FrameCaptureId captureId = AZ::Render::InvalidFrameCaptureId;
+
+        m_pipeline->AddToRenderTickOnce();
+        AZ::Render::FrameCaptureRequestBus::BroadcastResult(
+            captureId,
+            &AZ::Render::FrameCaptureRequestBus::Events::CapturePassAttachmentWithCallback,
+            m_passHierarchy,
+            AZStd::string("Output"),
+            callback,
+            AZ::RPI::PassAttachmentReadbackOption::Output);
+    }
+
+    const CameraSensorDescription& CameraSensor::GetCameraSensorDescription() const
+    {
+        return m_cameraSensorDescription;
+    }
+
+    void CameraSensor::publishMassage(
+        std::shared_ptr<rclcpp::Publisher<sensor_msgs::msg::Image>> publisher,
```

Comment:
I am assuming this is from the Robotics2 API and has to be a `std::shared_ptr`?

Reply:
Exactly, that pointer is given by rclcpp : https://docs.ros2.org/beta2/api/rclcpp/classrclcpp_1_1subscription_1_1Subscription.html#acce99fbedb205a9952cdd52ebff3ff45

Reply:
I confirm, we should keep std:: when dealing with ROS 2 API